typescript/patterns/src/xa_transactions.ts

```diff
@@ -0,0 +1,123 @@
+        let weRunTheQuery = false;
+        const txnId = await ctx.sideEffect(() => {
+            weRunTheQuery = true;
+            return Promise.resolve(randomUUID());
+        });
+
+        // step 2: if we actually ran that side effect to generate the txnId, (rather
+        //         than replay it) then we execute the query. If we didn't execute that
+        //         side effect, then we won't attempt commit that query anyways and will
+        //         instead vote to roll it back, so it doesn't matter if the query was
+        //         actually never run.
+        //         As an optimization, we don't put this into a sideEffect, because it
+        //         will anyways never re-execute on replay, because it is conditioned on
+        //         the side effect before.
+        let queryResult: ResultT | undefined = undefined;
+        if (weRunTheQuery) {
+
+            // In this block, cleanup of failures behaves like in any other app as well.
+            // for control-flow-style aborts, we abort the transaction. For hard failures,
+            // we rely in the DBMS to cancel the txn when the connection closes
+            // in case the transaction was already prepared when we failed, the following steps
+            // will take care of the cleanup
+            try {
+                await database.beginTxn();
+                queryResult = await database.runSql(query);
+                await database.prepareTxn(txnId);
+            } catch (e) {
+                await database.abortTxn();
+                throw e;
+            }
+        }
+
+        // step 3: we vote whether to abort or to accept the query. Only if we executed the
+        //         query before, we'll vote to commit it, otherwise we'll vote to abort it.
+        //         That way, we ensure that any recovery/replay for a previous execution that
+        //         didn't make it till the vote-to-commit step will make sure the possible query
+        //         run gets cleaned up in order to release DB txn locks.
+        // type CommittedResult = { commit: boolean, queryResult?: ResultT }
+
+        const commitDecision = await ctx.sideEffect(() =>
+                Promise.resolve({ commit: weRunTheQuery, queryResult }));
+
+        if (commitDecision.commit) {
+            // now we need to ensure this one gets committed
+            await ctx.sideEffect(() => database.commitPreparedTxn(txnId));
+            
+            // the result could come from a previous execution, in case that execution
+            // crashed just after committing (in Restate) the commit decision, but before
+            // signalling the database to commit (or returning the result)
+            return commitDecision.queryResult!;
+        } else {
+            // we clean up this query. it might be that this query was never prepared, but
+            // that does not matter here
+            await ctx.sideEffect(() => database.rollbackPreparedTxn(txnId));
```

Comment:
What happens in the following scenario?

Assume we have a first call that manages to persist its `txnId` and then before running the `weRunTheQuery` if block it pauses executing. Now Restate retries the invocation. The retry attempt will reach this point and rolls back the prepared txn which hasn't been created. Now the original invocation continues running and creates a prepared txn which acquires all the locks. If then the second invocation attempt enters the `weRunTheQuery` block and runs `runSql` what would happen? Would the `runSql` execution fail because it cannot acquire the locks or would it wait on acquiring the locks?